### PR TITLE
Share side modules with worker threads via postMessage

### PR DESCRIPTION
**Pull request by sbc100**
Any side modules that are loaded at the time of worker creation are         
shared with the worker via the initial postMessage.                                  
                                                                         
As a followup we should extend this to modules that are loaded after        
the worker is created but before the pthread runs (for example when         
a module is loaded while a worker is unused).                            
                                                                         
Fixes: #18552

### CODE REVIEW

**Comment by sbc100** [comment]
I've already split several PRs out from this one.. and I could split a couple more perhaps.  But PTAL and see what you think?

File: src/library_dylink.js [hunk 1]
```diff
@@ -359,7 +359,14 @@ var LibraryDylink = {
...
     // handle  -> dso; Used by dlsym
     loadedLibsByHandle: {},
-    init: () => newDSO('__main__', {{{ cDefs.RTLD_DEFAULT }}}, wasmImports),
+    init: () => {
+#if ASSERTIONS
+      // This function needs to run after the intial wasmImports object
+      // as been created.
+      assert(wasmImports);
+#endif
+      newDSO('__main__', {{{ cDefs.RTLD_DEFAULT }}}, wasmImports);
```
Comment (kripken) [inline src/library_dylink.js:368]: Do we not need a `return` here? There are `{ }` so I think we do?

File: src/library_dylink.js [hunk 2]
```diff
@@ -761,12 +782,11 @@ var LibraryDylink = {
         }
 #if STACK_OVERFLOW_CHECK >= 2
         if (moduleExports['__set_stack_limits']) {
-#if PTHREADS
           // When we are on an uninitialized pthread we delay calling
           // __set_stack_limits until $setDylinkStackLimits.
-          if (!ENVIRONMENT_IS_PTHREAD || runtimeInitialized)
-#endif
-          moduleExports['__set_stack_limits']({{{ to64('_emscripten_stack_get_base()') }}}, {{{ to64('_emscripten_stack_get_end()') }}});
+          if (runtimeInitialized) {
```
Comment (kripken) [inline src/library_dylink.js:787]: The comment says "on a pthread" but the check for that was removed?

File: emcc.py [hunk 3]
```diff
@@ -2076,7 +2076,10 @@ def phase_linker_setup(options, state, newargs):
     assert not settings.SIDE_MODULE
     if settings.MAIN_MODULE == 1:
       settings.INCLUDE_FULL_LIBRARY = 1
+    # Called from preamble.js once the main module is instantiated.
     settings.DEFAULT_LIBRARY_FUNCS_TO_INCLUDE += ['$loadDylibs']
+    if settings.STACK_OVERFLOW_CHECK == 2:
+      settings.DEFAULT_LIBRARY_FUNCS_TO_INCLUDE += ['$setDylinkStackLimits']
```
Comment (kripken) [inline emcc.py:2082]: Is this stack limit change related?

File: src/library_pthread.js [hunk 4]
```diff
@@ -105,6 +105,14 @@ var LibraryPThread = {
...
       }
 #endif
+#if !MINIMAL_RUNTIME
+      // MINIMAL_RUNTIME takes care of calling loadWasmModuleToAllWorkers
+      // in postamble_minimal.js
+      addOnPreRun(() => {
+        addRunDependency('loading-workers')
+        PThread.loadWasmModuleToAllWorkers(() => removeRunDependency('loading-workers'));
+      });
+#endif
```
Comment (kripken) [inline src/library_pthread.js:115]: Is this a general change to how startup is handled that could be in a separate PR?

File: src/postamble.js [hunk 5]
```diff
@@ -131,11 +131,10 @@ function stackCheckInit() {
...
 #endif
 
-#if RELOCATABLE
-var dylibsLoaded = false;
-#if '$LDSO' in addedLibraryItems
-LDSO.init();
-#endif
+#if MAIN_MODULE && PTHREADS
+// Map of modules to be shared with new threads.  This gets populated by the
+// main thread an shared with all new workers.
```
Comment (kripken) [inline src/postamble.js:136]:
```suggestion
// main thread and shared with all new workers.
```

File: src/postamble.js [hunk 6]
```diff
@@ -131,11 +131,10 @@ function stackCheckInit() {
...
+var sharedModules = Module['sharedModules'] ||[];
```
Comment (kripken) [inline src/postamble.js:137]: Does this really happen before `Module['sharedModules']` is set? But this is at the top level scope, so it happens immediately, while the shared modules arrive in a message..?

Comment (kripken) [inline src/postamble.js:137]:
```suggestion
var sharedModules = Module['sharedModules'] || [];
```

File: test/test_core.py [hunk 7]
```diff
@@ -9474,34 +9474,39 @@ def test_pthread_dylink_main_module_1(self):
...
-      }
+      Module['dynamicLibraries'] = ['liblib.so'];
     ''')
 
+    if args:
+      self.setup_node_pthreads()
+      create_file('post.js', '''
+        if (ENVIRONMENT_IS_PTHREAD) {
+          err('sharedModules: ' + Object.keys(sharedModules));
+          assert('liblib.so' in sharedModules);
```
Comment (kripken) [inline test/test_core.py:9499]: Maybe also assert on the type of the object in `sharedModules['liblib.so']`?

File: src/library_dylink.js (hunk 1 above)
Comment (sbc100) [reply src/library_dylink.js:368]: I'm not trying to return anything.. just setup the special DSO zero

Comment (kripken) [reply src/library_dylink.js:368]: Oh, I see, so the return value can be ignored here. sgtm.

File: test/test_core.py (hunk 7 above)
Comment (sbc100) [reply test/test_core.py:9499]: In JS things like WebAssembly just have `typeof(..) == 'object'` so its not very interesting to assert that.. but I could if you think its worth it?

File: src/postamble.js (hunk 6 above)
Comment (sbc100) [reply src/postamble.js]: Yes that way this works in worker.js is that we set a bunch of properties on the Worker object before importing the main script.  e.g.: https://github.com/emscripten-core/emscripten/blob/32043824e01ec6154b700bc410f30473fb277732/src/worker.js#L177

From the POV of the main script these act like incoming properties of the Module, just like if they had been set by a user in the non-threaded case.

File: test/test_core.py (hunk 7 above)
Comment (kripken) [reply test/test_core.py:9499]: Oh, I was thinking we could do `instanceof WebAssembly.Module` - is that not the case? Not a big deal if not, I guess.

File: src/postamble.js (hunk 6 above)
Comment (kripken) [reply src/postamble.js]: Ah, thanks, I forgot that stuff...

File: test/test_core.py (hunk 7 above)
Comment (sbc100) [reply test/test_core.py:9499]: Doh.. yes I realized that right after I send the last reply.  Fixed.

**Comment by sbc100** [comment]
Rebased on top of move_loadDylibs

File: emcc.py (hunk 3 above)
Comment (kripken) [reply emcc.py:2082]: This looks like part of the PR it's rebased on, but I see it isn't. Is it an old version of that PR's emcc changes maybe? Or why is it needed here?

**Review by kripken — APPROVED** [review]
Much easier to understand now, thanks!

lgtm % question
